### PR TITLE
feat(opensearch-vectorindex): replace docker-based custom resource with typescript

### LINKED ISSUE #567: Removing Docker build dependency from build phase
While creating OpenSearch serverless vector index, we’re using Dockerized python app to create mappings in given vector collection, CDK environments are mostly Docker based solutions and leads to docker-in-docker support requirements.

We can remove the Docker phrase from the initial configuration to make the process simpler ^^

https://github.com/awslabs/generative-ai-cdk-constructs/blob/fab3bbaadd54ef56f5a70e20f47111f403cbb53a/lambda/opensearch-serverless-custom-resources/custom_resources/opensearch_index.py#L179

### PR DESCRIPTION
**Pull request by cagataycali**
feat(opensearch-vectorindex): replace docker-based custom resource with typescript implementation

- remove docker dependency for opensearch vector index creation
- implement typescript-based lambda function for index management
- add support for configurable vector search parameters
- enhance vectorindex construct with new optional properties
- update custom resource provider to use bundled typescript code
- improve error handling and logging in lambda function
- wip: further testing and integration required

Fixes #567 #566

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

File: src/common/helpers/custom-resource-provider-helper.ts [hunk 1]
```diff
@@ -114,7 +114,7 @@ export function buildCustomResourceProvider(props: CRProviderProps): ICRProvider
       });
 
       const customResourceFunction = new lambda.Function(this, 'CustomResourcesFunction', {
-        code: lambda.Code.fromDockerBuild(codePath),
```
Comment (krokoko) [inline src/common/helpers/custom-resource-provider-helper.ts:117]: other constructs are using this helper method, and require to keep using the "fromDockerBuild" API. We can update the helper logic to support both use cases

Comment (krokoko) [reply src/common/helpers/custom-resource-provider-helper.ts:117]: Updated in latest commit

File: test/common/helpers/custom-resource-provider-helper.test.ts [hunk 2]
```diff
@@ -20,9 +20,9 @@ import { buildCustomResourceProvider } from '../../../src/common/helpers/custom-
 const AOSSCRProvider = buildCustomResourceProvider({
   providerName: 'OpenSearchIndexCRProvider',
   codePath: path.join(
-    __dirname, '../../../lambda/opensearch-serverless-custom-resources'),
-  handler: 'custom_resources.on_event',
-  runtime: lambda.Runtime.PYTHON_3_12,
+    __dirname, '../../../lambda/opensearch-serverless-custom-resources/index.ts'),
+  handler: 'handler',
+  runtime: lambda.Runtime.NODEJS_18_X,
```
Comment (scottschreckengaust) [inline test/common/helpers/custom-resource-provider-helper.test.ts:25]: Consider using latest AWS Lambda runtime?
```suggestion
  runtime: lambda.Runtime.NODEJS_20_X,
```

**Comment by krokoko** [comment]
Cannot get this to build and run. Tested a different approach by updating the custom-resource-provider-helper to use

```
const pathToBuildFile = codePath+'/build.js';
        const pathToOutputFile = codePath;
        const handler = 'index.handler';

        const commandThatIsRanDuringCdkSynth = ['node', pathToBuildFile];
        const code = lambda.Code.fromCustomCommand(
          pathToOutputFile,
          commandThatIsRanDuringCdkSynth,
        );
        
        customResourceFunction = new lambdanode.NodejsFunction(this, 'CustomResourcesFunction', {
          code: code,
          handler,
          runtime,
          role: this.role,
          timeout: cdk.Duration.minutes(15),
          memorySize: 128,
          vpc,
          vpcSubnets: vpc ? { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS } : undefined,
          securityGroups: vpc && securityGroup ? [securityGroup] : undefined,
          logRetention: logs.RetentionDays.ONE_WEEK,
          description: 'Custom Resource Provider',
        });
```

And have the original files in the folder to invoke the custom build script

**Comment by krokoko** [comment]
Update:
Locally it is building and passing tests. 
- Failing on mutation as the esbuild step is changing the content of index.js
- Tested with a sample stack to deploy a kb with oss and getting an error:
image

**Pull request closed without merging**